ubuntu/main.cpp

```diff
@@ -235,70 +198,71 @@ static void aa_touch_event(HU::TouchInfo::TOUCH_ACTION action, unsigned int x, u
-	    if (ret < 0) {
-	        printf("aa_touch_event(): hu_aap_enc_send() failed with (%d)\n", ret);
-	    }
-	});
+        {
+            HU::InputEvent inputEvent;
+            inputEvent.set_timestamp(get_cur_timestamp());
+            HU::TouchInfo* touchEvent = inputEvent.mutable_touch();
+            touchEvent->set_action(action);
+            HU::TouchInfo::Location* touchLocation = touchEvent->add_location();
+            touchLocation->set_x(x);
+            touchLocation->set_y(y);
+            touchLocation->set_pointer_id(0);
+
+            /* Send touch event */
+
+            int ret = s.hu_aap_enc_send_message(0, AA_CH_TOU, HU_INPUT_CHANNEL_MESSAGE::InputEvent, inputEvent);
+            if (ret < 0) {
+                printf("aa_touch_event(): hu_aap_enc_send() failed with (%d)\n", ret);
+            }
+        });
 }
 
 
 static void read_mic_data (GstElement * sink)
 {
-        
+    GstSample *gstsample;
     GstBuffer *gstbuf;
-    int ret;
-    
-    g_signal_emit_by_name (sink, "pull-buffer", &gstbuf,NULL);
     
+    gstsample = gst_app_sink_pull_sample((GstAppSink *)sink);
+    gstbuf = gst_sample_get_buffer(gstsample);
+
     if (gstbuf) {
 
-        /* if mic is stopped, don't bother sending */    
+        /* if mic is stopped, don't bother sending */
         GstState mic_state = GST_STATE_NULL;
         if (gst_element_get_state (mic_pipeline, &mic_state, NULL, GST_CLOCK_TIME_NONE) != GST_STATE_CHANGE_SUCCESS
-            || mic_state != GST_STATE_PLAYING) {
+                || mic_state != GST_STATE_PLAYING) {
             printf("Mic stopped.. dropping buffers \n");
             gst_buffer_unref(gstbuf);
             return;
         }
-        
-        
-        gint mic_buf_sz = GST_BUFFER_SIZE (gstbuf);
-        
-        int idx;
-        
-        if (mic_buf_sz <= 64) {
-            printf("Mic data < 64 \n");
-            gst_buffer_unref(gstbuf);
-            return;
+
+        GstMapInfo mapInfo;
+        if (gst_buffer_map(gstbuf, &mapInfo, GST_MAP_READ))
+        {
+
+            if (mapInfo.size <= 64) {
+                printf("Mic data < 64 \n");
+                gst_buffer_unref(gstbuf);
+                return;
+            }
+
+            uint64_t bufTimestamp = GST_BUFFER_TIMESTAMP(gstbuf);
+            uint64_t timestamp = GST_CLOCK_TIME_IS_VALID(bufTimestamp) ? (bufTimestamp / 1000) : get_cur_timestamp();
+            g_hu->hu_queue_command([timestamp, gstbuf, &mapInfo](IHUConnectionThreadInterface& s)
+            {
+                int ret = s.hu_aap_enc_send_media_packet(1, AA_CH_MIC, HU_PROTOCOL_MESSAGE::MediaDataWithTimestamp, timestamp, mapInfo.data, mapInfo.size);
+
+                if (ret < 0) {
+                    printf("read_mic_data(): hu_aap_enc_send() failed with (%d)\n", ret);
+                }
+
+                gst_buffer_unref(gstbuf);
+            });
+                gst_buffer_unmap(gstbuf, &mapInfo);
```

Comment:
Good catch there! Having a look at it again: gst_buffer_unmap needs to be before gst_buffer_unref, but they both need to be in the lambda, otherwise gstreamer will keep warning.